### PR TITLE
spec: add specification

### PR DESCRIPTION
**Pull request by banteg**
specify v0.3

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            - this party can terminate and clawback the escrow before it fully vests
+            - set to `empty(address)` to disable clawback, this can also be done later with `escrow`
+    - constraints
+        - `token`, `amount`
+            - `token` that doesn't return True on transfer is supported via `default_return_value=True`
+            - implicitly checked by `token` transfer
+                - the funder must have a `token` balance of at least `amount`
+                - the factory must have a `token` allowance of at least `amount`
+        - `vesting_start`
+            - can be in the past
```
Comment (storming0x) [inline spec.md:29]: i guess some check is needed if the vesting starts in the past and the length is over by blocktime, it just ends up being a transfer with extra steps right?

File: spec.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    - return an appropriate value or True from methods to make contract calls cheaper
+    - use `default_return_value=True` for token transfers
+
+- `__init__`
+    - actions
+        - template contract must be made defunct by storing `initialized` as True
+- `initialize`
+    - arguments
+        - token
+        - recipient
```
Comment (storming0x) [inline spec.md:51]: dont see it in master branch, but vesting previous version lacked a 0x0 check for recipient, i guess the admin can clawback in this edge case scenario right?

File: spec.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+- `deploy_vesting_escrow`
+    - arguments
+        - `token: address` ERC20 funding token
+        - `recipient: address` vesting recipient
+            - can be a smart contract in v0.3
+        - `amount: uint256` total vesting amount
+        - `vesting_duration: uint256` in seconds
+        - `vesting_start: uint256 = block.timestamp` in unix epoch
+        - `cliff_duration: uint256 = 0` in seconds
+        - `open_claim: bool = True`
```
Comment (storming0x) [inline spec.md:17]: immutable arg im guessing?

File: spec.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        - time must be now or in the future
+        - time must not exceed vesting end time
+    - actions
+        - `disabled_at` is set to `time`
+        - admin is set to `empty(address)`
+        - the amount of tokens is determined as tokens still locked at `time`
+        - tokens are transferred to `admin`
+        - log `VestingTerminated(self.recipient, self.admin, rugged, time)`
+        - log `SetFree(self.admin)`
+- `set_free` (previously `set_admin`)
```
Comment (storming0x) [inline spec.md:97]: usually i seen `renounce_admin` or `renounce_ownership` used elsewhere to clarify the intention of the func.

set_free may be understood as a setting called `free` at first glance.

File: spec.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        - set admin to `empty(address)`
+        - log `SetFree(self.admin)`
+- `collect_dust`
+    - arguments
+        - `token: address` dust token to claim
+        - `beneficiary: address = msg.sender` where to send tokens to
+    - constraints
+        - recipient can claim to any beneficiary
+        - if beneficiary is the recipient and `open_claim` is True, allow anyone to claim
+        - if the token is the vested token itself, the amount is determined as token balance of the contract minus the still locked portion of the vesting
```
Comment (storming0x) [inline spec.md:112]: awesome addition

File: spec.md (hunk 3 above)
Comment (banteg) [reply spec.md:17]: it's the initial value that can be changed by `recipient`. we use this default to be able to create vestings towards contracts. then anyone would be able to claim it for them.

File: spec.md (hunk 1 above)
Comment (banteg) [reply spec.md:29]: maybe we can assert that `vesting_start + vesting_duration <= block.timestamp`, but i don't see anything that could break if someone creates a fully vested contract.

File: spec.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        - recipient can claim to any beneficiary
+        - if beneficiary is the recipient and `open_claim` is True, allow anyone to claim
+        - the `amount` is limited by the maximum amount claimable
+    - actions
+        - `total_claimed` is updated
+        - transfer tokens to beneficiary, use `default_return_value=True`
+        - log `Claim` with beneficiary and claimed amount
+    - returns
+        - amount claimed
+- `terminate` (previously `rug_pull`)
```
[the remark this answers is not in the document]
Comment (banteg) [reply spec.md:81]: no, even in the previous version, `rug_pull` didn't bite into the already vested portion, we didn't change the math so it's safe.

File: spec.md (hunk 2 above)
Comment (banteg) [reply spec.md:56]: added

File: spec.md (hunk 4 above)
Comment (banteg) [reply spec.md:102]: will change to `disown``

File: spec.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+## Vesting Escrow
+
+- global considerations
+    - return an appropriate value or True from methods to make contract calls cheaper
+    - use `default_return_value=True` for token transfers
+
+- `__init__`
+    - actions
+        - template contract must be made defunct by storing `initialized` as True
```
Comment (kadmil) [inline spec.md:57]: Just in case, the architecture doesn't use blueprints, that right?

File: spec.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        - amount claimed
+- `revoke` (previously `rug_pull`)
+    - arguments
+        - `time: uint256 = block.timestamp` optionally terminate at a date and clawback lower amount
+        - `beneficiary: address` where to send the clawed back amount
+            - one use case could be to reward the recipient early, another is to transfer to treasury and not have owner handle to funds
+    - constraints
+        - can only be called by `owner`
+        - time must be now or in the future
+        - time must not exceed vesting end time
```
Comment (kadmil) [inline spec.md:99]: clawback only non-vested amount?

File: spec.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        - `support_vyper: uint256 = 100` in bps
+    - constraints
+        - `token`, `amount`
+            - `token` that doesn't return True on transfer is supported via `default_return_value=True`
+            - implicitly checked by `token` transfer
+                - the funder must have a `token` balance of at least `amount`
+                - the factory must have a `token` allowance of at least `amount`
+        - `vesting_start + vesting_duration`
+            - end time must be in the future
+            - a regular transfer would suit better if it's in the past
```
[the remark this answers is not in the document]
Comment (banteg) [reply spec.md:31]: i agree, and we allow start time in the past. we just decided to disallow the whole vest being in the past. it doesn't make sense and can help catch wrong usage early.

File: spec.md (hunk 7 above)
Comment (banteg) [reply spec.md:57]: yes, after looking at usage data, we decided against blueprint and reverted back to minimal proxies.

**Review by milkyklim — APPROVED** [review]

**Pull request merged**